### ISSUE TITLE
runtime differences with Eigen 3.3-rc1 + patch

### ISSUE THREAD

**Issue by bgoodri** [stan-dev/math#187]
The coming 3.3 release of Eigen has more changes than usual
http://eigen.tuxfamily.org/index.php?title=3.3
And Stan Math has a few build errors with it.

```
./stan/math/prim/mat/fun/LDLT_factor.hpp:104:24: error: no type named 'solve_retval' in namespace 'Eigen::internal'
./stan/math/prim/mat/fun/LDLT_factor.hpp:104:36: error: expected member name or ';' after declaration specifiers
./stan/math/rev/mat/fun/Eigen_NumTraits.hpp:127:66: error: template argument for template type parameter must be a type
./stan/math/rev/mat/fun/Eigen_NumTraits.hpp:157:42: error: template argument for template type parameter must be a type
./stan/math/rev/mat/fun/Eigen_NumTraits.hpp:201:59: error: template argument for template type parameter must be a type
./stan/math/rev/mat/fun/LDLT_factor.hpp:85:24: error: no type named 'solve_retval' in namespace 'Eigen::internal'
./stan/math/rev/mat/fun/LDLT_factor.hpp:85:36: error: expected member name or ';' after declaration specifiers
```

There was a bug fix in a branch that's not a release yet that builds, but gives wrong answers as @syclik indicated in https://github.com/stan-dev/math/issues/187#issuecomment-254578730

**Comment by bob-carpenter** [comment]
It would be great to fix this in such a way it works with the current Eigen and the forthcoming 3.3 release.

I took a look at the traits, but this is going to take some digging.  Are those multiply operations in the NumTraits file really going to be used internally by Eigen for multiplication?  I didn't quite follow what they were doing there or which interface they were implementing.

**Comment by bgoodri** [comment]
Now Eigen is in beta. Math is still failing.
```
goodrich@T540p:/opt/stan/lib/stan_math_2.7.0$ grep -F "error:" /tmp/eigen_test.txt | sort | uniq
./stan/math/prim/mat/fun/LDLT_factor.hpp:104:24: error: no type named 'solve_retval' in namespace 'Eigen::internal'
./stan/math/prim/mat/fun/LDLT_factor.hpp:104:36: error: expected member name or ';' after declaration specifiers
./stan/math/rev/mat/fun/Eigen_NumTraits.hpp:128:66: error: template argument for template type parameter must be a In file included from <built-in>:325:
./stan/math/rev/mat/fun/Eigen_NumTraits.hpp:128:66: error: template argument for template type parameter must be a type
./stan/math/rev/mat/fun/Eigen_NumTraits.hpp:128:66: error:          typename LhsScalar, typename LhsMapper, int LhsStorageOrder, bool ConjugateLhs,
./stan/math/rev/mat/fun/Eigen_NumTraits.hpp:158:42: error: template argument for template typeIn file included from  parameter must be a type
./stan/math/rev/mat/fun/Eigen_NumTraits.hpp:158:42: error: template argument for template type parameter must be a type
./stan/math/rev/mat/fun/Eigen_NumTraits.hpp:202:59: error: template argument for template type parameter must be a type
./stan/math/rev/mat/fun/LDLT_factor.hpp:85:24: error: no type named 'solve_retval' in namespace 'Eigen::internal'
./stan/math/rev/mat/fun/LDLT_factor.hpp:85:36: error: expected member name or ';' after declaration specifiers
```

**Comment by bob-carpenter** [comment]
RcppEigen is still at 3.2.8, as of today.  I'm going to update to that and see if I can get it working with the 3.3 beta, but if not, I'll just update to 3.2.8, for which I'm going to create a new issue.

**Comment by bgoodri** [comment]
3.2.8 should be fine for now. But 3.3.0 is going to introduce a lot of
changes, so we need to test it a lot before it is released, which we can't
do if it doesn't build anything Eigen-related.

**Comment by bgoodri** [comment]
Still not building with beta2; now worse
```
./stan/math/fwd/mat/fun/Eigen_NumTraits.hpp:95:12: error: ‘significant_decimals_default_impl’ is not a class template
./stan/math/fwd/mat/fun/Eigen_NumTraits.hpp:95:72: error: ‘Eigen::internal::significant_decimals_default_impl’ is not a template
./stan/math/prim/arr/fun/fill.hpp:23:12: error: ‘size_t’ was not declared in this scope
./stan/math/prim/arr/fun/fill.hpp:23:26: error: ‘i’ was not declared in this scope
./stan/math/prim/mat/fun/LDLT_factor.hpp:104:24: error: ‘solve_retval’ in namespace ‘Eigen::internal’ does not name a template type
./stan/math/rev/mat/fun/Eigen_NumTraits.hpp:130:54: error: wrong number of template arguments (6, should be at least 8)
./stan/math/rev/mat/fun/Eigen_NumTraits.hpp:159:71: error: wrong number of template arguments (6, should be at least 8)
./stan/math/rev/mat/fun/Eigen_NumTraits.hpp:192:69: error: ‘ReturnType’ in ‘struct Eigen::internal::scalar_product_traits<stan::math::var, stan::math::var>’ does not name a type
./stan/math/rev/mat/fun/Eigen_NumTraits.hpp:203:78: error: wrong number of template arguments (6, should be at least 8)
./stan/math/rev/mat/fun/Eigen_NumTraits.hpp:210:46: error: expression list treated as compound expression in initializer [-fpermissive]
./stan/math/rev/mat/fun/Eigen_NumTraits.hpp:96:12: error: ‘significant_decimals_default_impl’ is not a class template
./stan/math/rev/mat/fun/Eigen_NumTraits.hpp:96:70: error: explicit specialization of non-template ‘Eigen::internal::significant_decimals_default_impl’
./stan/math/rev/mat/fun/LDLT_factor.hpp:85:24: error: ‘solve_retval’ in namespace ‘Eigen::internal’ does not name a template type
```

**Comment by syclik** [comment]
Looks like we need to change how we define traits for Eigen. Some of it we can remove. Some of it we'll have to rewrite.

**Comment by bob-carpenter** [comment]
Is there a way to check version and create ifdef so it works
for both 3.2 and 3.3?

**Comment by syclik** [comment]
Good question.

I've been digging into it and most of the errors we can handle generally
for both 3.2 and 3.3. When we can't, we'll need to find some way to switch
on version.

**Comment by bob-carpenter** [comment]
We're only tied to supporting what RcppEigen does.  So we could
put the updates on a branch, and then switch to that when RcppEigen
updates.  We control which version of Eigen is used for math
and CmdStan (and presumably PyStan, where we don't have to work
around CRAN's restrictive size policies).

- Bob

**Comment by bob-carpenter** [comment]
We're good to go for Eigen 3.3 now.

**Comment by bgoodri** [comment]
with which commit?

**Comment by bgoodri** [comment]
If this was fixed, it never got merged.

**Comment by bob-carpenter** [comment]
@syclik : is this still on a branch?  Or was there something you didn't manage to get building?  I thought we had this sorted out.

**Comment by syclik** [comment]
No. I don't think there's a branch. Perhaps we never got it fully working
-- I don't know. What can we do to test?


Daniel

**Comment by bgoodri** [comment]
We can do what we did before: run `make -k test-headers` and observe the error messages:
```
note: 'significant_decimals_impl' declared ./stan/math/rev/mat/fun/Eigen_NumTraits.hpp:146:59: error: template argument for template type parameter must be a type
./stan/math/fwd/mat/fun/Eigen_NumTraits.hpp:40:12: error: no template named 'significant_decimals_default_impl'; did you mean 'significant_decimals_impl'?
./stan/math/fwd/mat/fun/Eigen_NumTraits.hpp:40:12: error: too many template arguments for class template 'significant_decimals_impl'
./stan/math/prim/arr/fun/fill.hpp:23:12: error: unknown type name 'size_t'; did you mean '__gnu_cxx::size_t'?
./stan/math/prim/mat/fun/LDLT_factor.hpp:95:24: error: no type named 'solve_retval' in namespace 'Eigen::internal'
./stan/math/prim/mat/fun/LDLT_factor.hpp:95:36: error: expected member name or ';' after declaration specifiers
./stan/math/prim/mat/fun/round.hpp:19:20: error: no member named 'round' in namespace 'std'
./stan/math/rev/mat/fun/Eigen_NumTraits.hpp:102:42: error: template argument for template type parameter must be a type
./stan/math/rev/mat/fun/Eigen_NumTraits.hpp:136:69: error: no type named 'ReturnType' in 'Eigen::internal::scalar_product_traits<stan::math::var, stan::math::var>'
./stan/math/rev/mat/fun/Eigen_NumTraits.hpp:146:59: error: template argument for template type parameter must be a type
./stan/math/rev/mat/fun/Eigen_NumTraits.hpp:40:12: error: no template named 'significant_decimals_default_impl'; did youmake/tests:78: recipe for target 'stan/math/rev/mat/fun/unit_vector_constrain.hpp-test' failed
./stan/math/rev/mat/fun/Eigen_NumTraits.hpp:40:12: error: no template named 'significant_decimals_default_impl'; did you mean 'significant_decimals_impl'?
./stan/math/rev/mat/fun/Eigen_NumTraits.hpp:40:12: error: too many template arguments for class template 'significant_decimals_impl'
./stan/math/rev/mat/fun/Eigen_NumTraits.hpp:40:12: /usr/include/eigen3/Eigen/src/Core/util/BlasUtil.h:38:39: note: error: no template named 'significant_decimals_default_impl'; did you mean 'significant_decimals_impl'?
./stan/math/rev/mat/fun/Eigen_NumTraits.hpp:72:66: error: template argument for template type parameter must be a type
./stan/math/rev/mat/fun/Eigen_NumTraits.hpp:72:66: error: template argument for template type parameter must make/tests:78: recipe for target 'stan/math/rev/scal/fun/primitive_value.hpp-test' failed
./stan/math/rev/mat/fun/Eigen_NumTraits.hpp:72:./stan/math/rev/mat/fun/Eigen_NumTraits.hpp:146:59: error: template argument for template type parameter must be a type
 ./stan/math/rev/mat/fun/Eigen_NumTraits.hpp:note102: :42'significant_decimals_impl':  declared hereerror: 
./stan/math/rev/mat/fun/LDLT_factor.hpp:86:24: error: no type named 'solve_retval' in namespace 'Eigen::internal'
./stan/math/rev/mat/fun/LDLT_factor.hpp:86:36: error: expected member name or ';' after declaration specifiers
```

**Comment by bob-carpenter** [comment]
You had created the requisite traits.  I'm not sure they were
compatible with Eigen 3.2, so we have a fork in the road.  I
hope we don't need to support both at any given time.  Can we
just wait until RcppEigen goes to 3.3 and switch then?

- Bob

**Comment by syclik** [comment]
This test currently fails:
```
> ./runTests.py test/unit/math/rev/mat/fun/cholesky_decompose_test.cpp 
------------------------------------------------------------
make test/unit/math/rev/mat/fun/cholesky_decompose_test
make: `test/unit/math/rev/mat/fun/cholesky_decompose_test' is up to date.
------------------------------------------------------------
test/unit/math/rev/mat/fun/cholesky_decompose_test --gtest_output="xml:test/unit/math/rev/mat/fun/cholesky_decompose_test.xml"
Running main() from gtest_main.cc
[==========] Running 3 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 3 tests from AgradRevMatrix
[ RUN      ] AgradRevMatrix.mat_cholesky
test/unit/math/rev/mat/fun/cholesky_decompose_test.cpp:183: Failure
Value of: LL_trans(0,0).val()
  Actual: inf
Expected: a.val()
Which is: 3
test/unit/math/rev/mat/fun/cholesky_decompose_test.cpp:184: Failure
Value of: LL_trans(0,1).val()
  Actual: -inf
Expected: b.val()
Which is: -1
test/unit/math/rev/mat/fun/cholesky_decompose_test.cpp:185: Failure
Value of: LL_trans(1,0).val()
  Actual: -inf
Expected: c.val()
Which is: -1
[  FAILED  ] AgradRevMatrix.mat_cholesky (0 ms)
[ RUN      ] AgradRevMatrix.exception_mat_cholesky
[       OK ] AgradRevMatrix.exception_mat_cholesky (0 ms)
[ RUN      ] AgradRevMatrix.mat_cholesky_1st_deriv
[       OK ] AgradRevMatrix.mat_cholesky_1st_deriv (538 ms)
[----------] 3 tests from AgradRevMatrix (538 ms total)

[----------] Global test environment tear-down
[==========] 3 tests from 1 test case ran. (538 ms total)
[  PASSED  ] 2 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] AgradRevMatrix.mat_cholesky

 1 FAILED TEST
test/unit/math/rev/mat/fun/cholesky_decompose_test --gtest_output="xml:test/unit/math/rev/mat/fun/cholesky_decompose_test.xml" failed
exit now (10/12/16 23:08:12 EDT)
```

**Comment by syclik** [comment]
After merging with develop, it looks like we have compilation issues:
```
./runTests.py test/unit/math/fwd/mat/fun/cholesky_decompose_test.cpp 
```

doesn't compile:
```
In file included from test/unit/math/fwd/mat/fun/cholesky_decompose_test.cpp:1:
In file included from ./stan/math/fwd/mat.hpp:8:
In file included from ./stan/math/fwd/mat/vectorize/apply_scalar_unary.hpp:4:
In file included from ./stan/math/prim/mat/vectorize/apply_scalar_unary.hpp:4:
In file included from lib/eigen_3.3.ffa8356/Eigen/Dense:3:
In file included from lib/eigen_3.3.ffa8356/Eigen/Cholesky:31:
lib/eigen_3.3.ffa8356/Eigen/src/Cholesky/LLT.h:321:45: error: use of overloaded operator '*' is
      ambiguous (with operand types 'Block<Eigen::Matrix<stan::math::fvar<double>, -1, -1, 0, -1,
      -1>, Dynamic, Dynamic>' and 'const AdjointReturnType' (aka 'const Eigen::Transpose<const
      Eigen::Block<Eigen::Matrix<stan::math::fvar<double>, -1, -1, 0, -1, -1>, 1, -1, false> >'))
      if (k>0 && rs>0) A21.noalias() -= A20 * A10.adjoint();
                                        ~~~ ^ ~~~~~~~~~~~~~
lib/eigen_3.3.ffa8356/Eigen/src/Cholesky/LLT.h:333:14: note: in instantiation of function template
      specialization 'Eigen::internal::llt_inplace<stan::math::fvar<double>,
      1>::unblocked<Eigen::Matrix<stan::math::fvar<double>, -1, -1, 0, -1, -1> >' requested here
      return unblocked(m);
             ^
lib/eigen_3.3.ffa8356/Eigen/src/Cholesky/LLT.h:397:61: note: in instantiation of function template
      specialization 'Eigen::internal::llt_inplace<stan::math::fvar<double>,
      1>::blocked<Eigen::Matrix<stan::math::fvar<double>, -1, -1, 0, -1, -1> >' requested here
  { return llt_inplace<typename MatrixType::Scalar, Lower>::blocked(m)==-1; }
                                                            ^
lib/eigen_3.3.ffa8356/Eigen/src/Cholesky/LLT.h:444:21: note: in instantiation of member function
      'Eigen::internal::LLT_Traits<Eigen::Matrix<stan::math::fvar<double>, -1, -1, 0, -1, -1>,
      1>::inplace_decomposition' requested here
  bool ok = Traits::inplace_decomposition(m_matrix);
                    ^
./stan/math/prim/mat/fun/cholesky_decompose.hpp:31:11: note: in instantiation of function template
      specialization 'Eigen::LLT<Eigen::Matrix<stan::math::fvar<double>, -1, -1, 0, -1, -1>,
      1>::compute<Eigen::Matrix<stan::math::fvar<double>, -1, -1, 0, -1, -1> >' requested here
      llt.compute(m);
          ^
test/unit/math/fwd/mat/fun/cholesky_decompose_test.cpp:40:28: note: in instantiation of function
      template specialization 'stan::math::cholesky_decompose<stan::math::fvar<double> >' requested
      here
  EXPECT_THROW(stan::math::cholesky_decompose(m),std::domain_error);
                           ^
lib/eigen_3.3.ffa8356/Eigen/src/SparseCore/../plugins/CommonCwiseBinaryOps.h:50:29: note: candidate
      function [with T = Eigen::Block<Eigen::Matrix<stan::math::fvar<double>, -1, -1, 0, -1, -1>, -1,
      -1, false>]
EIGEN_MAKE_SCALAR_BINARY_OP(operator*,product)
                            ^
lib/eigen_3.3.ffa8356/Eigen/src/SparseCore/../plugins/CommonCwiseBinaryOps.h:50:29: note: candidate
      function [with T = Eigen::Transpose<const Eigen::Block<Eigen::Matrix<stan::math::fvar<double>,
      -1, -1, 0, -1, -1>, 1, -1, false> >]
lib/eigen_3.3.ffa8356/Eigen/src/Core/MatrixBase.h:173:5: note: candidate function [with OtherDerived
      = Eigen::Transpose<const Eigen::Block<Eigen::Matrix<stan::math::fvar<double>, -1, -1, 0, -1,
      -1>, 1, -1, false> >]
    operator*(const MatrixBase<OtherDerived> &other) const;
    ^
...
```

**Comment by bob-carpenter** [comment]
I think that's another one we just have to report to Eigen.  I don't see how we could fix it on our side as it's just a straight-up call to `llt.compute` on an `fvar<double>` that leads to ambiguity between their binary componentwise sparse ops:

```
Eigen/src/SparseCore/../plugins/CommonCwiseBinaryOps.h:50:29
```

and their dense base op:

```
Core/MatrixBase.h:173:5
```

**Comment by syclik** [comment]
What's odd is that I think it worked before the merge with develop. I don't
see how, but that's what I remember.

I didn't have a chance to actually hunt down what's going on. I'll put
together a minimal working example and submit it to Eigen when I have a
free minute.

**Comment by bob-carpenter** [comment]
OK.  Sounds good.  I have zero suggestions on how to make
a simple example.  I assume it works with double type.

- Bob

**Comment by bob-carpenter** [comment]
This is still not working.  Eigen dug us into a pretty deep hole with their reorganized traits.  I'm still trying to dig our code out.

**Comment by syclik** [comment]
I just updated the branch with Eigen 3.3-rc2. I haven't checked behavior yet. (Hopefully you don't mind... if you need, just back out the last commit on that branch.)

**Comment by bob-carpenter** [comment]
Thanks---that's great.  I have all of prim, rev, and fwd working,
but now need to figure out include orders for mix.  I had to
write a new autodiff-promotable traits metaprogram in order to
for the enable_if to control the fvar constructor.   I'll send word
when I re-emerge from the rabbit hole.

- Bob

**Comment by syclik** [comment]
Update: the branch isn't working for reverse mode's matrix inverse function. This will fail now:
`./runTests.py test/unit/math/rev/mat/fun/inverse_test.cpp`

With a long hairy message that starts with:
```
In file included from test/unit/math/rev/mat/fun/inverse_test.cpp:1:
In file included from ./stan/math/rev/mat.hpp:4:
In file included from ./stan/math/rev/core.hpp:14:
In file included from ./stan/math/rev/core/matrix_vari.hpp:4:
In file included from ./stan/math/rev/mat/fun/Eigen_NumTraits.hpp:4:
In file included from ./stan/math/prim/mat/fun/Eigen.hpp:4:
In file included from lib/eigen_3.3-rc2/Eigen/Dense:1:
In file included from lib/eigen_3.3-rc2/Eigen/Core:480:
lib/eigen_3.3-rc2/Eigen/src/Core/products/GeneralMatrixMatrix.h:233:26: error: no type named 'Traits' in 'Eigen::internal::general_matrix_matrix_product<long, stan::math::var, 0, false, stan::math::var,
      0, false, 0>'
  typedef typename Gemm::Traits Traits;
          ~~~~~~~~~~~~~~~^~~~~~
lib/eigen_3.3-rc2/Eigen/src/Core/products/GeneralMatrixMatrix.h:484:10: note: in instantiation of template class 'Eigen::internal::gemm_functor<stan::math::var, long,
      Eigen::internal::general_matrix_matrix_product<long, stan::math::var, 0, false, stan::math::var, 0, false, 0>, Eigen::Block<Eigen::Block<Eigen::Map<Eigen::Matrix<stan::math::var, -1, -1, 0, -1, -1>,
      0, Eigen::Stride<0, 0> >, -1, -1, false>, -1, -1, false>, Eigen::Block<Eigen::Block<Eigen::Map<Eigen::Matrix<stan::math::var, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> >, -1, -1, false>, -1, -1,
      false>, Eigen::Block<Eigen::Block<Eigen::Map<Eigen::Matrix<stan::math::var, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> >, -1, -1, false>, -1, -1, false>, Eigen::internal::gemm_blocking_space<0,
      stan::math::var, stan::math::var, -1, -1, -1, 1, false> >' requested here
        (GemmFunctor(lhs, rhs, dst, actualAlpha, blocking), a_lhs.rows(), a_rhs.cols(), a_lhs.cols(), Dest::Flags&RowMajorBit);
         ^
...
```

I just looked at the diff of the Eigen 3.3-rc2 code with Eigen 3.2.9 and it's very different.

**Comment by bob-carpenter** [comment]
They completely rewrote all their expression templates!

- Bob

**Comment by syclik** [comment]
I'm starting to learn that. =). It looks like they even changed from rc1 to rc2.

The issue I'm having now is that `stan/math/rev/mat/Eigen_NumTraits.hpp` doesn't match the base templates defined by Eigen. I'm not sure how `run(...)` works with `LhsMapper` and `LhsScalar` and the relationship between them. I need to dig some more.

@bgoodri, do you know how to implement this off the top of your head?

**Comment by bgoodri** [comment]
no

**Comment by syclik** [comment]
ok. back to reading undocumented code.

**Comment by bgoodri** [comment]
My guess is that their AutoDiff module has to do the same thing(s)

https://bitbucket.org/eigen/eigen/src/42149289d06757fbdaf49d750d948dc91ea751c5/unsupported/Eigen/src/AutoDiff/?at=default

**Comment by bob-carpenter** [comment]
I see what happened---I reverted some of my changes.  If you
comment out all the code in namespace internal in 

/stan/math/rev/mat/fun/Eigen_NumTraits.hpp

then that test (and I think all of rev) passes.  Not so sure
because I'm at an airport now and have to get on a plane before
the tests could finish.

**Comment by syclik** [comment]
I commented out the bit in the `Eigen::internal` namespace and fixed an argument dependent lookup problem in forward mode's determinant. I'm running tests now. It looks ok.

We can try to figure out how to properly deal with the template specializations for `stan::math::var` tomorrow.

**Comment by syclik** [comment]
All tests passed. We're compiling and running ok again.

**Comment by bob-carpenter** [comment]
I added the rest of the NumTraits enums based on their built-in autodiff example in Eigen's unsupported directory.  The mixed type traits were OK, so I left those in.

What needed to be commented out was our template specializations for the structs for `general_matrix_vector_product` and `general_matrix_matrix_product` in terms of our `gevv_vvv_vari` reverse-mode variable implementation.  

 This may cause efficiency issues in situations where we autodiff through complex matrix operations.

**Comment by bob-carpenter** [comment]
I think I have the overloads figured out in such a way that the code will work with 3.3 and earlier versions with our reverse-mode specializations.  They encapsulated a few of the arguments in larger structs, so I just had to tear them apart.

**Comment by bob-carpenter** [comment]
I added back in the matrix-vector and matrix-matrix product specializations and the necessary traits.  Everything is passing now with Eigen 3.3.0.  I think this should be bacward compatible with older versions of Eigen, but haven't tested.

**Comment by bgoodri** [comment]
@bob-carpenter Do you have a more definite conclusion now about whether the next Stan will be compatible with both versions of Eigen?

**Comment by syclik** [comment]
@bgoodri, do we need to be able to work with both Eigen 3.2 and Eigen 3.3?

**Comment by bgoodri** [comment]
That would be ideal. In that scenario, we can do a 2.13 release of
StanHeaders / rstan / rstanarm and then whenever RcppEigen updates to
3.3.x, we won't have to do anything. But I think RcppEigen is ready to
update to 3.3.x now essentially, so if we can't be compatible with both
3.2.x and 3.3.x, it would be better to go with 3.3.x.

**Comment by bob-carpenter** [comment]
It'll probably take a few hours of work to get a version compatible
with both.  All we need to do is extreme #if-ing on the version
numbers for the things that are different.  That'll also make it easy
to clean up later.  Give us a couple hours.

Then we're ready to release 2.13.  The matrix exponential still hasn't
passed tests, but we could do a 2.13 release without that if it's urgent.
And without updating the manual, for that matter.

- Bob

**Comment by syclik** [comment]
For us to support both 3.2 and 3.3, we're going to need to branch a fair
amount using C macros. If it's not necessary, then we can just switch when
it's released. If it is, then we'll put in the time now to get the
branching in place. It's not as simple as we first thought.

**Comment by bgoodri** [comment]
We need https://github.com/stan-dev/stan/pull/2109 before the 2.13 release. I'll check again on the schedule for RcppEigen.

**Comment by bob-carpenter** [comment]
I just reviewed it.  Just waiting on the tests to pass and it
can be merged.

- Bob

**Comment by bob-carpenter** [comment]
This one's been dealt with.
